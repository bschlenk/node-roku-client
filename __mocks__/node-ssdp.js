const ssdp = jest.genMockFromModule('node-ssdp');

let headers = null;

class Client {
  /**
   * Construct a mock ssdp client.
   */
  constructor() {
    this.searched = null;
    this.events = {};
  }

  /**
   * Mock .search()
   * @param {string} query 'ssdp:all'
   */
  search(query) {
    this.searched = query;
    if (!headers) {
      return;
    }
    const callResponse = (index = null) => {
      const { response } = this.events;
      if (response) {
        if (index === null) {
          response(headers);
        } else {
          response(headers[index]);
        }
      }
    };
    if (Array.isArray(headers)) {
      let index = 0;
      const { length } = headers;
      setTimeout(function recurseResponses() {
        callResponse(index);
        index += 1;
        if (index < length) {
          setTimeout(recurseResponses, 0);
        }
      }, 0);
    } else {
      setTimeout(callResponse, 0);
    }
  }

  /**
   * Mock .on() event listener method
   * @param {string}   eventName 'response'
   * @param {function} callback pass data back to callee
   * @return {{ SERVER: string, LOCATION: string }}
   */
  on(event, fn) {
    this.events[event] = fn;
  }

  stop() {}
}

ssdp.Client = Client;

ssdp.__setHeaders = function __setHeaders(h) {
  headers = h;
};

module.exports = ssdp;
