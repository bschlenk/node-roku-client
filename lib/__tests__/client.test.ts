import * as fs from 'fs';
import * as path from 'path';
import * as stream from 'stream';
import Client, { DEFAULT_PORT } from '../client';
import fetchPonyfill = require('fetch-ponyfill');
import { FetchMock } from 'jest-fetch-mock';

const fetchObjects = fetchPonyfill();
const fetch = fetchObjects.fetch as FetchMock;

const clientAddr = 'http://192.168.1.61:8060';

function loadResponse(name: string): string;
function loadResponse(name: string, asBuffer: false): string;
function loadResponse(name: string, asBuffer: true): ReadableStream;
function loadResponse(name: string, asBuffer = false) {
  const file = name.includes('.') ? name : `${name}.xml`;
  const data = fs.readFileSync(path.join(__dirname, 'assets', file));
  if (!asBuffer) {
    return data.toString('utf-8');
  }
  const bufferStream = new stream.PassThrough();
  bufferStream.end(data);
  return (bufferStream as unknown) as ReadableStream;
}

describe('Client', () => {
  let client: Client;

  beforeEach(() => {
    client = new Client(clientAddr);
    fetch.mockClear();
  });

  describe('#discover()', () => {
    it('should resolve to a client instance for the first address found', () => {
      require('node-ssdp').__setHeaders({
        SERVER: 'Roku UPnP/1.0 MiniUPnPd/1.4',
        LOCATION: 'http://192.168.1.17:8060/dial/dd.xml',
      });
      return Client.discover().then((c) => {
        expect(c).toBeInstanceOf(Client);
        expect(c.ip).toEqual('http://192.168.1.17:8060');
      });
    });
  });

  describe('#constructor()', () => {
    it('should construct a new Client object', () => {
      expect(client).toBeDefined();
      expect(client.ip).toEqual(clientAddr);
    });

    it('should add missing http://', () => {
      const c = new Client('192.168.1.2:1111');
      expect(c.ip).toEqual('http://192.168.1.2:1111');
    });

    it('should add the default port if omitted', () => {
      const c = new Client('192.168.1.2');
      expect(c.ip).toEqual(`http://192.168.1.2:${DEFAULT_PORT}`);
    });
  });

  describe('#apps()', () => {
    it('should return a list of apps', () => {
      fetch.mockResponse(loadResponse('apps'));
      return client.apps().then((apps) => {
        expect(apps).toBeInstanceOf(Array);
        apps.forEach((app) => {
          expect(app).toEqual(
            expect.objectContaining({
              id: expect.any(String),
              name: expect.any(String),
              type: expect.any(String),
              version: expect.any(String),
            }),
          );
        });
      });
    });
  });

  describe('#active()', () => {
    it('should return the active app', () => {
      fetch.mockResponse(loadResponse('active-app'));
      return client.active().then((app) => {
        expect(app).toEqual(
          expect.objectContaining({
            id: expect.any(String),
            name: expect.any(String),
            type: expect.any(String),
            version: expect.any(String),
          }),
        );
      });
    });

    it('should return null if there is not an active app', () => {
      fetch.mockResponse(loadResponse('active-app-none'));
      return client.active().then((app) => {
        expect(app).toBeNull();
      });
    });

    it('should reject if multiple apps are returned', () => {
      fetch.mockResponse(loadResponse('active-multiple'));
      return client
        .active()
        .then(() => {
          throw new Error('Should have thrown');
        })
        .catch((err) => {
          expect(err).toBeDefined();
          expect(err).toBeInstanceOf(Error);
        });
    });
  });

  describe('#info()', () => {
    it('should return info for the roku device', () => {
      fetch.mockResponse(loadResponse('info'));
      return client.info().then((info) => {
        expect(info).toBeInstanceOf(Object);
        expect(Object.keys(info).length).toEqual(29);
        expect(info['model-name']).toBeUndefined();
        expect(info.modelName).toEqual('Roku 3');
      });
    });
  });

  describe('#keypress()', () => {
    it('should press the home button', () =>
      client.keypress('Home').then(() => {
        expect(fetch).toHaveBeenCalledWith(`${clientAddr}/keypress/Home`, {
          method: 'POST',
        });
      }));

    it('should send a Lit_ command if a single character is passed in', () => {
      client.keypress('a').then(() => {
        expect(fetch).toHaveBeenCalledWith(`${clientAddr}/keypress/Lit_a`, {
          method: 'POST',
        });
      });
    });

    it('should url encode Lit_ commands for utf-8 characters', () => {
      client.keypress('â‚¬').then(() => {
        expect(fetch).toHaveBeenCalledWith(
          `${clientAddr}/keypress/Lit_%E2%82%AC`,
          {
            method: 'POST',
          },
        );
      });
    });
  });

  describe('#keydown()', () => {
    it('should press and hold the pause', () =>
      client.keydown('Pause').then(() => {
        expect(fetch).toHaveBeenCalledWith(`${clientAddr}/keydown/Pause`, {
          method: 'POST',
        });
      }));
  });

  describe('#keyup()', () => {
    it('should release the info button', () =>
      client.keyup('Info').then(() => {
        expect(fetch).toHaveBeenCalledWith(`${clientAddr}/keyup/Info`, {
          method: 'POST',
        });
      }));
  });

  describe('#icon()', () => {
    it('should download the icon to the given folder', () => {
      const response = new fetchObjects.Response(
        loadResponse('netflix.jpeg', true),
        { headers: new fetchObjects.Headers({ 'content-type': 'image/jpeg' }) },
      );
      fetch.mockImplementation(() => Promise.resolve(response));
      return client.icon('12').then((icon) => {
        expect(icon.type).toEqual('image/jpeg');
        expect(icon.extension).toEqual('.jpeg');
        expect(icon.response).toBe(response);
      });
    });
  });

  describe('#launch()', () => {
    it('should call launch for the given app id', () =>
      client.launch('12345').then(() => {
        expect(fetch).toHaveBeenCalledWith(`${client.ip}/launch/12345`, {
          method: 'POST',
        });
      }));
  });

  describe('#launchDtv()', () => {
    it('should call launch/tvinput.dtv', () => {
      client.launchDtv().then(() => {
        expect(fetch).toHaveBeenCalledWith(`${client.ip}/launch/tvinput.dtv`, {
          method: 'POST',
        });
      });
    });

    it('should pass a channel string to launch', () => {
      client.launchDtv('1.1').then(() => {
        expect(fetch).toHaveBeenCalledWith(
          `${client.ip}/launch/tvinput.dtv?ch=1.1`,
          {
            method: 'POST',
          },
        );
      });
    });

    it('should pass a channel number to launch', () => {
      client.launchDtv(8.5).then(() => {
        expect(fetch).toHaveBeenCalledWith(
          `${client.ip}/launch/tvinput.dtv?ch=8.5`,
          {
            method: 'POST',
          },
        );
      });
    });
  });

  describe('#text()', () => {
    it('should send a Lit_ command for each letter', () =>
      client.text('hello').then(() => {
        expect(fetch.mock.calls).toEqual([
          [`${client.ip}/keypress/Lit_h`, { method: 'POST' }],
          [`${client.ip}/keypress/Lit_e`, { method: 'POST' }],
          [`${client.ip}/keypress/Lit_l`, { method: 'POST' }],
          [`${client.ip}/keypress/Lit_l`, { method: 'POST' }],
          [`${client.ip}/keypress/Lit_o`, { method: 'POST' }],
        ]);
      }));
  });

  describe('#command()', () => {
    it('should allow chaining remote commands', () =>
      client
        .command()
        .volumeUp()
        .select()
        .text('abc')
        .send()
        .then(() => {
          expect(fetch.mock.calls).toEqual([
            [`${client.ip}/keypress/VolumeUp`, { method: 'POST' }],
            [`${client.ip}/keypress/Select`, { method: 'POST' }],
            [`${client.ip}/keypress/Lit_a`, { method: 'POST' }],
            [`${client.ip}/keypress/Lit_b`, { method: 'POST' }],
            [`${client.ip}/keypress/Lit_c`, { method: 'POST' }],
          ]);
        }));
  });
});
