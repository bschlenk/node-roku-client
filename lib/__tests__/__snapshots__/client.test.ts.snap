// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client #info() should return info for the roku device 1`] = `
Object {
  "advertisingId": "ec7f71ee-35d7-5a2b-bafd-e4b9f3947337",
  "country": "US",
  "developerEnabled": false,
  "deviceId": "4E652H070911",
  "ethernetMac": "ac:3a:7a:80:ae:08",
  "headphonesConnected": false,
  "keyedDeveloperId": "",
  "language": "en",
  "locale": "en_US",
  "modelName": "Roku 3",
  "modelNumber": "4230X",
  "modelRegion": "US",
  "networkType": "wifi",
  "notificationsEnabled": true,
  "notificationsFirstUse": true,
  "powerMode": "PowerOn",
  "searchEnabled": true,
  "secureDevice": true,
  "serialNumber": "4E652H070911",
  "softwareBuild": "4100",
  "softwareVersion": "7.2.0",
  "supportsSuspend": false,
  "timeZone": "US/Hawaii",
  "timeZoneOffset": "-600",
  "udn": "04546142-6001-1014-80ff-ac3a7a80ae09",
  "userDeviceName": "",
  "vendorName": "Roku",
  "voiceSearchEnabled": true,
  "wifiMac": "ac:3a:7a:80:ae:09",
}
`;

exports[`Client #mediaPlayer() should parse the response 1`] = `
Object {
  "buffering": Object {
    "current": "1000",
    "max": "1000",
    "target": "0",
  },
  "duration": "887999 ms",
  "error": false,
  "format": Object {
    "audio": "aac",
    "captions": "none",
    "container": "mp4",
    "drm": "none",
    "video": "mpeg4_15",
    "videoRes": "1280x546",
  },
  "isLive": false,
  "newStream": Object {
    "speed": "128000 bps",
  },
  "plugin": Object {
    "bandwidth": "44692475 bps",
    "id": "dev",
    "name": "MultiLive",
  },
  "position": "6916 ms",
  "runtime": "887999 ms",
  "state": "play",
  "streamSegment": Object {
    "bitrate": "0",
    "mediaSequence": "1",
    "segmentType": "mux",
    "time": "0",
  },
}
`;
